src/controls/tab-bar/focus.js

```diff
@@ -0,0 +1,93 @@
+// List of key codes that can be used to modify the DOM's activeElement in a tab list
+const PERTINENT_KEY_CODES = {
+  DOWN: '40',
+  UP: '38',
+  HOME: '36',
+  END: '35',
+  LEFT: '37',
+  RIGHT: '39',
+}
+
+// Recursively searches for the closest parent tab list
+function getClosestTabList (el) {
+  return el && (el.matches('[role="tablist"]') ? el : getClosestTabList(el.parentElement))
+}
+
+// Checks if the element is a tab
+function isTabControl (el) {
+  return el && el.matches('[role="tab"]')
+}
+
+// Moves focus the next tab in the tab list
+// Note: this will wrap around if the current tab is the first or last element
+function focusNextControl (control, direction) {
+  const tabList = getClosestTabList(control)
+  const tabs = Array.from(tabList.querySelectorAll('[role="tab"]'))
+  const currentIndex = tabs.indexOf(control)
+  
+  if (currentIndex === -1) return
+  
+  const newIndex = getSafeIndex(tabs.length, currentIndex + 1*direction)
+  tabs[newIndex].focus()
+}
+
+// Moves focus to the "extreme" tab (first or last in the tab list)
+function focusExtremeControl (control, top=true) {
+  const tabList = getClosestTabList(control)
+  const tabs = Array.from(tabList.querySelectorAll('[role="tab"]'))
+  
+  top ? tabs[0].focus() : tabs[tabs.length - 1].focus()
+}
+
+// Recursively calculates a "safe" index by wrapping around when the array bounds are exceeded
+// e.g., getSafeIndex([1, 2, 3].length, 3) -> 0
+function getSafeIndex (length, newIndex) {
```

Comment:
Can use the modulus operator for this! `safeIndex = index % array.length`

Reply:
Does that handle the case where the index is negative? e.g., `fn([0, 1, 2], -1) -> 2`

Reply:
Looks like you need a floored modulo - so a helper is still necessary ☹️ https://stackoverflow.com/a/43827557

Reply:
🤯

Reply:
By the way, ruby (and lots of other languages) would do that by default. Also I found this neat function too (it doesn't wrap though): https://lodash.com/docs/4.17.15#nth